File: javascript/speech/speaker-recognition.js
// <dependencies>
"use strict";

/* To run this sample, install:
npm install microsoft-cognitiveservices-speech-sdk
*/
var sdk = require("microsoft-cognitiveservices-speech-sdk");
var fs = require("fs");

// Note: Change the locale if desired.
const profile_locale = "en-us";

/* Note: passphrase_files and verify_file should contain paths to audio files that contain \"My voice is my passport, verify me.\"
You can obtain these files from:
https://github.com/Azure-Samples/cognitive-services-speech-sdk/tree/fa6428a0837779cbeae172688e0286625e340942/quickstart/javascript/node/speaker-recognition/verification
*/ 
const passphrase_files = ["myVoiceIsMyPassportVerifyMe01.wav", "myVoiceIsMyPassportVerifyMe02.wav", "myVoiceIsMyPassportVerifyMe03.wav"];
const verify_file = "myVoiceIsMyPassportVerifyMe04.wav";
/* Note: identify_file should contain a path to an audio file that uses the same voice as the other files, but contains different speech. You can obtain this file from:
https://github.com/Azure-Samples/cognitive-services-speech-sdk/tree/fa6428a0837779cbeae172688e0286625e340942/quickstart/javascript/node/speaker-recognition/identification
*/
const identify_file = "aboutSpeechSdk.wav";

var subscription_key = 'PASTE_YOUR_SPEECH_SUBSCRIPTION_KEY_HERE';
var region = 'PASTE_YOUR_SPEECH_ENDPOINT_REGION_HERE';

const ticks_per_second = 10000000;
// </dependencies>

// <helpers>
/* From: https://github.com/Azure-Samples/cognitive-services-speech-sdk/blob/fa6428a0837779cbeae172688e0286625e340942/quickstart/javascript/node/speaker-recognition/verification/dependent-verification.js#L8
*/
function GetAudioConfigFromFile (file)
{
	return sdk.AudioConfig.fromWavFileInput(fs.readFileSync(file));
}
// </helpers>

// <get_activation_phrases>
async function GetActivationPhrases(client, type, locale)
{
	try {
		console.log ("Getting Activation Phrases for enrollment");
		const result = await client.getActivationPhrasesAsync(type, locale);
		if (!!result.phrases && result.phrases.length > 0) {
			for (const phrase of result.phrases) {
				console.log (`Activation phrase received: ${phrase}\n`);
			}
		} else {
			console.log ("No activation phrases received.\n");
		}
	} catch (error) {
		console.log ("Error getting activation phrases: " + error);
	}
}
// </get_activation_phrases>

// <add_enrollments_dependent>
async function AddEnrollmentsToTextDependentProfile(client, profile, audio_files)
{
	try {
		for (const file of audio_files) {
			console.log ("Adding enrollment to text dependent profile...");
			const audio_config = GetAudioConfigFromFile(file);
			const result = await client.enrollProfileAsync(profile, audio_config);
			if (result.reason === sdk.ResultReason.Canceled) {
				throw(JSON.stringify(sdk.VoiceProfileEnrollmentCancellationDetails.fromResult(result)));
			}
			else {
				console.log ("Remaining enrollments needed: " + result.privDetails["remainingEnrollmentsCount"] + ".");
			}
		};
		console.log ("Enrollment completed.\n");
	} catch (error) {
		console.log ("Error adding enrollments: " + error);
	}
}
// </add_enrollments_dependent>

// <add_enrollments_independent>
async function AddEnrollmentsToTextIndependentProfile(client, profile, audio_files)
{
	try {
		for (const file of audio_files) {
			console.log ("Adding enrollment to text independent profile...");
			const audio_config = GetAudioConfigFromFile(file);
			const result = await client.enrollProfileAsync (profile, audio_config);
			if (result.reason === sdk.ResultReason.Canceled) {
				throw(JSON.stringify(sdk.VoiceProfileEnrollmentCancellationDetails.fromResult(result)));
			}
			else {
				console.log ("Remaining audio time needed: " + (result.privDetails["remainingEnrollmentsSpeechLength"] / ticks_per_second) + " seconds.");
			}
		}
		console.log ("Enrollment completed.\n");
	} catch (error) {
		console.log ("Error adding enrollments: " + error);
	}
}
// </add_enrollments_independent>

// <speaker_verify>
async function SpeakerVerify(profile, recognizer)
{
	try {
		const model = sdk.SpeakerVerificationModel.fromProfile(profile);
		const result = await recognizer.recognizeOnceAsync(model);
		console.log ("Verified voice profile for speaker: " + result.profileId + ". Score is: " + result.score + ".\n");
	} catch (error) {
		console.log ("Error verifying speaker: " + error);
	}
}
// </speaker_verify>

// <speaker_identify>
async function SpeakerIdentify(profile, recognizer)
{
	try {
		const model = sdk.SpeakerIdentificationModel.fromProfiles([profile]);
		const result = await recognizer.recognizeOnceAsync(model);
		console.log ("The most similar voice profile is: " + result.profileId + " with similarity score: " + result.score + ".\n");
	} catch (error) {
		console.log ("Error identifying speaker: " + error);
	}
}
// </speaker_identify>

// <text_dependent_verification>
async function TextDependentVerification(client, speech_config)
{
	console.log ("Text Dependent Verification:\n");
	var profile = null;
	try {
		const type = sdk.VoiceProfileType.TextDependentVerification;
		// Create the profile.
		profile = await client.createProfileAsync(type, profile_locale);
		console.log ("Created profile ID: " + profile.profileId);
		// Get the activation phrases
		await GetActivationPhrases(type, profile_locale);
		await AddEnrollmentsToTextDependentProfile(client, profile, passphrase_files);
		const audio_config = GetAudioConfigFromFile(verify_file);
		const recognizer = new sdk.SpeakerRecognizer(speech_config, audio_config);
		await SpeakerVerify(profile, recognizer);
	}
	catch (error) {
		console.log ("Error:\n" + error);
	}
	finally {
		if (profile !== null) {
			console.log ("Deleting profile ID: " + profile.profileId);
			const deleteResult = await client.deleteProfileAsync (profile);
		}
	}
}
// </text_dependent_verification>

// <text_independent_verification>
async function TextIndependentVerification(client, speech_config)
{
	console.log ("Text Independent Verification:\n");
	var profile = null;
	try {
		const type = sdk.VoiceProfileType.TextIndependentVerification;
		// Create the profile.
		profile = await client.createProfileAsync(type, profile_locale);
		console.log ("Created profile ID: " + profile.profileId);
		// Get the activation phrases
		await GetActivationPhrases(type, profile_locale);
		await AddEnrollmentsToTextIndependentProfile(client, profile, [identify_file]);
		const audio_config = GetAudioConfigFromFile(passphrase_files[0]);
		const recognizer = new sdk.SpeakerRecognizer(speech_config, audio_config);
		await SpeakerVerify(profile, recognizer);
	}
	catch (error) {
		console.log ("Error:\n" + error);
	}
	finally {
		if (profile !== null) {
			console.log ("Deleting profile ID: " + profile.profileId);
			const deleteResult = await client.deleteProfileAsync (profile);
		}
	}
}
// </text_independent_verification>

// <text_independent_indentification>
async function TextIndependentIdentification(client, speech_config)
{
	console.log ("Text Independent Identification:\n");
	var profile = null;
	try {
		const type = sdk.VoiceProfileType.TextIndependentIdentification;
		// Create the profile.
		profile = await client.createProfileAsync(type, profile_locale);
		console.log ("Created profile ID: " + profile.profileId);
		// Get the activation phrases
		await GetActivationPhrases(type, profile_locale);
		await AddEnrollmentsToTextIndependentProfile(client, profile, [identify_file]);
		const audio_config = GetAudioConfigFromFile(passphrase_files[0]);
		const recognizer = new sdk.SpeakerRecognizer(speech_config, audio_config);
		await SpeakerIdentify(profile, recognizer);
	}
	catch (error) {
		console.log ("Error:\n" + error);
	}
	finally {
		if (profile !== null) {
			console.log ("Deleting profile ID: " + profile.profileId);
			const deleteResult = await client.deleteProfileAsync (profile);
		}
	}
}
// </text_independent_indentification>

// <main>
async function main() {
	const speech_config = sdk.SpeechConfig.fromSubscription(subscription_key, region);
	const client = new sdk.VoiceProfileClient(speech_config);

	await TextDependentVerification(client, speech_config);
	await TextIndependentVerification(client, speech_config);
	await TextIndependentIdentification(client, speech_config);
	console.log ("End of quickstart.");
}
main();
// </main>

File: cpp/speech/speaker-recognition.cpp
// <dependencies>
#include <iostream>
#include <stdexcept>
// Note: Install the NuGet package Microsoft.CognitiveServices.Speech.
#include <speechapi_cxx.h>

using namespace std;
using namespace Microsoft::CognitiveServices::Speech;

// Note: Change the locale if desired.
auto profile_locale = "en-us";
auto audio_config = Audio::AudioConfig::FromDefaultMicrophoneInput();
auto ticks_per_second = 10000000;
// </dependencies>

// <get_speech_config>
shared_ptr<SpeechConfig> GetSpeechConfig()
{
	auto subscription_key = 'PASTE_YOUR_SPEECH_SUBSCRIPTION_KEY_HERE';
	auto region = 'PASTE_YOUR_SPEECH_ENDPOINT_REGION_HERE';
	auto config = SpeechConfig::FromSubscription(subscription_key, region);
	return config;
}
// </get_speech_config>

// <add_enrollments_dependent>
void AddEnrollmentsToTextDependentProfile(shared_ptr<VoiceProfileClient> client, shared_ptr<VoiceProfile> profile)
{
	shared_ptr<VoiceProfileEnrollmentResult> enroll_result = nullptr;
	auto phraseResult = client->GetActivationPhrasesAsync(profile->GetType(), profile_locale).get();
    auto phrases = phraseResult->GetPhrases();
	while (enroll_result == nullptr || enroll_result->GetEnrollmentInfo(EnrollmentInfoType::RemainingEnrollmentsCount) > 0)
	{
		if (phrases != nullptr && phrases->size() > 0)
		{
			std::cout << "Please say the passphrase, \"" << phrases->at(0) << "\"\n";
			enroll_result = client->EnrollProfileAsync(profile, audio_config).get();
			std::cout << "Remaining enrollments needed: " << enroll_result->GetEnrollmentInfo(EnrollmentInfoType::RemainingEnrollmentsCount) << ".\n";
		}
		else
		{
			std::cout << "No passphrases received, enrollment not attempted.\n\n";
		}
	}
	std::cout << "Enrollment completed.\n\n";
}
// </add_enrollments_dependent>

// <add_enrollments_independent>
void AddEnrollmentsToTextIndependentProfile(shared_ptr<VoiceProfileClient> client, shared_ptr<VoiceProfile> profile)
{
	shared_ptr<VoiceProfileEnrollmentResult> enroll_result = nullptr;
	auto phraseResult = client->GetActivationPhrasesAsync(profile->GetType(), profile_locale).get();
    auto phrases = phraseResult->GetPhrases();
	while (enroll_result == nullptr || enroll_result->GetEnrollmentInfo(EnrollmentInfoType::RemainingEnrollmentsSpeechLength) > 0)
	{
		if (phrases != nullptr && phrases->size() > 0)
		{
			std::cout << "Please say the activation phrase, \"" << phrases->at(0) << "\"\n";
			enroll_result = client->EnrollProfileAsync(profile, audio_config).get();
			std::cout << "Remaining audio time needed: " << enroll_result->GetEnrollmentInfo(EnrollmentInfoType::RemainingEnrollmentsSpeechLength) / ticks_per_second << " seconds.\n";
		}
		else
		{
			std::cout << "No activation phrases received, enrollment not attempted.\n\n";
		}
	}
	std::cout << "Enrollment completed.\n\n";
}
// </add_enrollments_independent>

// <speaker_verify>
void SpeakerVerify(shared_ptr<VoiceProfile> profile, shared_ptr<SpeakerRecognizer> recognizer)
{
	shared_ptr<SpeakerVerificationModel> model = SpeakerVerificationModel::FromProfile(profile);
	std::cout << "Speak the passphrase to verify: \"My voice is my passport, verify me.\"\n";
	shared_ptr<SpeakerRecognitionResult> result = recognizer->RecognizeOnceAsync(model).get();
	std::cout << "Verified voice profile for speaker: " << result->ProfileId << ". Score is: " << result->GetScore() << ".\n\n";
}
// </speaker_verify>

// <speaker_identify>
void SpeakerIdentify(shared_ptr<VoiceProfile> profile, shared_ptr<SpeakerRecognizer> recognizer)
{
	shared_ptr<SpeakerIdentificationModel> model = SpeakerIdentificationModel::FromProfiles({ profile });
	// Note: We need at least four seconds of audio after pauses are subtracted.
	std::cout << "Please speak for at least ten seconds to identify who it is from your list of enrolled speakers.\n";
	shared_ptr<SpeakerRecognitionResult> result = recognizer->RecognizeOnceAsync(model).get();
	std::cout << "The most similar voice profile is: " << result->ProfileId << " with similarity score: " << result->GetScore() << ".\n\n";
}
// </speaker_identify>

// <text_dependent_verification>
void TextDependentVerification(shared_ptr<VoiceProfileClient> client, shared_ptr<SpeakerRecognizer> recognizer)
{
	std::cout << "Text Dependent Verification:\n\n";
	// Create the profile.
	auto profile = client->CreateProfileAsync(VoiceProfileType::TextDependentVerification, profile_locale).get();
	std::cout << "Created profile ID: " << profile->GetId() << "\n";
	AddEnrollmentsToTextDependentProfile(client, profile);
	SpeakerVerify(profile, recognizer);
	// Delete the profile.
	client->DeleteProfileAsync(profile);
}
// </text_dependent_verification>

// <text_independent_verification>
void TextIndependentVerification(shared_ptr<VoiceProfileClient> client, shared_ptr<SpeakerRecognizer> recognizer)
{
	std::cout << "Text Independent Verification:\n\n";
	// Create the profile.
	auto profile = client->CreateProfileAsync(VoiceProfileType::TextIndependentVerification, profile_locale).get();
	std::cout << "Created profile ID: " << profile->GetId() << "\n";
	AddEnrollmentsToTextIndependentProfile(client, profile);
	SpeakerVerify(profile, recognizer);
	// Delete the profile.
	client->DeleteProfileAsync(profile);
}
// </text_independent_verification>

// <text_independent_indentification>
void TextIndependentIdentification(shared_ptr<VoiceProfileClient> client, shared_ptr<SpeakerRecognizer> recognizer)
{
	std::cout << "Speaker Identification:\n\n";
	// Create the profile.
	auto profile = client->CreateProfileAsync(VoiceProfileType::TextIndependentIdentification, profile_locale).get();
	std::cout << "Created profile ID: " << profile->GetId() << "\n";
	AddEnrollmentsToTextIndependentProfile(client, profile);
	SpeakerIdentify(profile, recognizer);
	// Delete the profile.
	client->DeleteProfileAsync(profile);
}
// </text_independent_indentification>

// <main>
int main()
{
	auto speech_config = GetSpeechConfig();
	auto client = VoiceProfileClient::FromConfig(speech_config);
	auto recognizer = SpeakerRecognizer::FromConfig(speech_config, audio_config);
	TextDependentVerification(client, recognizer);
	TextIndependentVerification(client, recognizer);
	TextIndependentIdentification(client, recognizer);
    std::cout << "End of quickstart.\n";
}
// </main>
